Lib/asyncio/tasks.py

```diff
@@ -43,22 +44,25 @@ def all_tasks(loop=None):
     """Return a set of all tasks for the loop."""
     if loop is None:
         loop = events.get_running_loop()
-    # Looping over a WeakSet (_all_tasks) isn't safe as it can be updated from another
-    # thread while we do so. Therefore we cast it to list prior to filtering. The list
-    # cast itself requires iteration, so we repeat it several times ignoring
-    # RuntimeErrors (which are not very likely to occur). See issues 34970 and 36607 for
-    # details.
+    # Looping over these sets isn't safe as they can be updated from another thread,
+    # therefore we cast to lists prior to filtering. The list cast itself requires
+    # iteration, so we repeat it several times ignoring RuntimeErrors (which are not
+    # very likely to occur). See issues 34970 and 36607 for details.
+    scheduled_tasks = None
+    eager_tasks = None
     i = 0
     while True:
         try:
-            tasks = list(_all_tasks)
+            if scheduled_tasks is None:
+                scheduled_tasks = list(_scheduled_tasks)
+            eager_tasks = list(_eager_tasks)
```

Comment:
Any particular reason we avoid re-casting `_scheduled_tasks`, but don't bother with that check for `_eager_tasks`?

Reply:
Yes, makes sense that the "other thread modifies" problem is only an issue for `WeakSet` and not for `set` (which should be thread-safe.) Given this, I think the comment above here shouldn't be modified as the PR currently modifies it to suggest that it's an issue for both.

Pulling the case of `_eager_tasks` outside the loop makes sense. Do we even have to cast it to list at all?

Reply:
Updated code looks good to me, modulo one tiny grammar nit